op-batcher/batcher/driver.go

```diff
@@ -464,20 +424,34 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
 				continue
 			}
 
-			l.state.pruneSafeBlocks(syncStatus.SafeL2)
-			l.state.pruneChannels(syncStatus.SafeL2)
+			// Decide appropriate actions
+			syncActions, outOfSync := computeSyncActions(*syncStatus, l.prevCurrentL1, l.state.blocks, l.state.channelQueue, l.Log)
 
-			err = l.state.CheckExpectedProgress(*syncStatus)
-			if err != nil {
-				l.Log.Warn("error checking expected progress, clearing state and waiting for node sync", "err", err)
-				l.waitNodeSyncAndClearState()
+			if outOfSync {
+				// If the sequencer is out of sync
+				// do nothing and wait to see if it has
+				// got in sync on the next tick.
+				l.Log.Warn("Sequencer is out of sync, retrying next tick.")
 				continue
 			}
 
-			if err := l.loadBlocksIntoState(*syncStatus, l.shutdownCtx); errors.Is(err, ErrReorg) {
-				l.Log.Warn("error loading blocks, clearing state and waiting for node sync", "err", err)
-				l.waitNodeSyncAndClearState()
-				continue
+			l.prevCurrentL1 = syncStatus.CurrentL1
+
+			// Manage existing state / garbage collection
+			if syncActions.clearState != nil {
+				l.state.Clear(*syncActions.clearState)
+			} else {
+				l.state.pruneSafeBlocks(syncActions.blocksToPrune)
+				l.state.pruneChannels(syncActions.channelsToPrune)
+			}
+
+			if syncActions.blocksToLoad != nil {
+				// Get fresh unsafe blocks
+				if err := l.loadBlocksIntoState(l.shutdownCtx, syncActions.blocksToLoad.start, syncActions.blocksToLoad.end); errors.Is(err, ErrReorg) {
+					l.Log.Warn("error loading blocks, clearing state and waiting for node sync", "err", err)
+					l.waitNodeSyncAndClearState()
```

Comment:
I am wondering if we want to keep this alternative way of "waiting for sync" and "clearing state:. 

Above, we are now waiting for sync purely in the sense that the sync status needs to be non empty and the currentl1 needs to not have reversed. Here, we wait for transactions to be ingested by the sequencer.

Above, we clear the `channelManager` state with the safel2.l1origin that we have from the sync status already in scope. Here, we go ahead and query for that again before clearing in the same way. 

Feels like an opportunity to harmonize, even delete some code?

Reply:
It's a good observation, definitely looks like something could be harmonized here. 

What makes this non-trivial is that we cannot detect the reorg case with 100% certainty earlier inside `computeSyncActions`. There might be a reorg between the time the block range is determined and then when we're actually fetching the blocks. So we have to check for reorgs at the time when we're actually processing the fetches blocks by calling `channelManager.AddL2Block` and see whether it returns an `ErrReorg`.

Maybe there's an opportunity to let the next `computeSyncActions` step handle the reorg in the next loop iteration, instead of handling it right there. But certainly needs some thought. I suggest we reserve this for a follow up and get this in as it is now.

Reply:
https://github.com/ethereum-optimism/optimism/issues/13279

Reply:
tracking: https://github.com/ethereum-optimism/optimism/issues/13279